Add custom logic for etags to all index actions

robbevp opened the pull request:
This adds custom logic for etag to our most used controllers, so that we can reduce the amount of work if the server has to do. Since browsers already keep a cache of requests, this should provide a big speed-up for web users (if they are requesting unchanged data). I'm not sure if the other clients will automatically benefit from this, or need to add logic to handle this themselves.

## Logic for stale data
We use rails' caching mechanism to calculate an etag for the whole collection. Internally `.cache_key_with_version` gets called on the scope. This looks at the SQL query that would be executed, the total number of items in the scope, the max updated_at in the collection. This means that the etag will be different if an item is added, updated or deleted.
The etag also includes the page number/amount of items per page, so two different requests will always include a different etag. (I'm not sure if this is strictly necessary, but it shouldn't hurt either?)

We only provide the `etag` header and don't provide a `Last-Modified` header, since that date wouldn't change if an item is deleted.

I added `touch: true` to the association that are included in the response, to avoid situations were we would update the associated object without changing `updated_at` on the "primary" object.

## Performance
This provides a big speed up for the controllers if the right `If-None-Match` header is send, since we don't need to fetch the actual data from the DB, and don't need to serialize the objects to json.
For most of these controllers the time spent by rails is reduced to ±10ms (on my local machine, in some casual observations - the real speed here will probably depend on the amount of records in the collection). Since a full reload of our data often sends a torrent of simultaneous requests, the duration measured by the client can still be a lot more (but that is no different from today).

## Further improvements
Since the etag is based on the total collection in each controller, a client could add the following logic to reduce the amount of work they have to do:
* Store the etag of the first page
* Request the first page with that etag
* Stop requesting pages if that first page matches

We could also base the etag on the actual page of items instead of the total collection, but that would mean that clients would still need to request all pages to verify that an item wasn't deleted somewhere. The current approach means that one change in one object would invalidate all pages. Since we often sort by `id: :desc` adding an item would always invalidate all pages (and deleting an item all pages from the one that contained that item onwards)

## Questions
* Do we want to add this to all index actions for consistency? I now only added this to the controllers which we used the most and often need to send multiple pages of data. I didn't add this to all index actions since we were missing timestamps on a lot of these models (no clue why we don't have those columns for every almost all models)
* Do we want more tests for this? I only added this for `AlbumsController` as an example, but I wouldn't mind adding this for all controllers that were changed. (Or adding tests to verify that changing some association results in a different etag)

- [x] I've added tests relevant to my changes.

<!---
  If you did any manual testing as well, please mention so
  here. Provide instructions so we can reproduce those tests.
  --->

chvp commented on app/controllers/plays_controller.rb:
```diff
@@ -6,10 +6,12 @@ class PlaysController < ApplicationController
+    @plays = scoped_plays.paginate(page: params[:page], per_page: params[:per_page])
     add_pagination_headers(@plays)
 
+    return unless stale?(etag: [scoped_plays.cache_key, scoped_plays.cache_version(:played_at)])
```
Why not add the timestamps for plays as well so that we can keep the logic the same everywhere? I honestly don't know anymore what the motivation was for leaving off the timestamps on some models; probably some misguided idea about the efficiency of not storing data that's not needed.

chvp reviewed (commented):
Nice! I definitely fundamentally agree with this PR. A few notes:
- I would add the check to all controllers for consistency like you said.
- If a `MAX(updated_at)` is executed to calculate the staleness, would it make sense to add indexes on that column to our tables? For larger tables like tracks and plays that could probably become slow without one.
- Same for the count of items. Since we're scoping both through policy and filtering, we should probably make sure that sufficient indexes exist to make that efficient.

chvp commented on app/controllers/plays_controller.rb:
```diff
@@ -6,10 +6,12 @@ class PlaysController < ApplicationController
+    @plays = scoped_plays.paginate(page: params[:page], per_page: params[:per_page])
     add_pagination_headers(@plays)
 
+    return unless stale?(etag: [scoped_plays.cache_key, scoped_plays.cache_version(:played_at)])
```
Right, forgot about the pagination headers. Not sure how that is handled in browser either (and definitely not sure about the other clients). But then why is it different for plays than all the other controllers?

robbevp replied:
There isn't a meaningful difference - because etag key here had to be constructed manually, it just felt more readable than:

```ruby
render json: @plays.map { transform_play_for_json(it) } if stale?(etag: [scoped_plays.cache_key, scoped_plays.cache_version(:played_at)])
```

But if plays also has an `updated_at` column, this will now just be:
```ruby
render json: @plays.map { transform_play_for_json(it) } if stale?(etag: scoped_plays)

chvp commented:
Thanks for the investigation. To me it seems worth it to calculate the etag manually to avoid that double count and get that max updated_at efficiently. When adding the indexes we should make sure to scope them by the columns used in the policy_scope.

robbevp commented:
@chvp This is ready for another look. A quick summary of the changes:
* Added an index for `updated_at` for the biggest table
* Added our own `stale?` method to calculate the etags manually. We do have an extra cached query to get the count (otherwise `stale?` would have to receive both the scope and the paginated collection)
  ```
    Play Count (13.9ms)  SELECT COUNT(*) FROM "plays" WHERE "plays"."user_id" = 1 /*action='index',application='Accentor',controller='plays'*/
  ↳ app/controllers/application_controller.rb:26:in 'ApplicationController#add_pagination_headers'
  Play Maximum (0.2ms)  SELECT MAX("plays"."updated_at") FROM "plays" WHERE "plays"."user_id" = 1 /*action='index',application='Accentor',controller='plays'*/
  ↳ app/controllers/application_controller.rb:39:in 'ApplicationController#stale?'
  CACHE Play Count (0.0ms)  SELECT COUNT(*) FROM "plays" WHERE "plays"."user_id" = 1
  ↳ app/controllers/application_controller.rb:40:in 'ApplicationController#stale?'
  ```
* My first draft wasn't very consistent in what was part of the scope we take into account for the etag. I now always include everything except the pagination (though the pagination params are still added to this in a different way)
* I used `reorder` in all places where we override `sorted`. This reduces the query from `SELECT "genres".* FROM "genres" ORDER BY "genres"."id" DESC, "genres"."id" ASC` to `SELECT "genres".* FROM "genres" ORDER BY "genres"."id" ASC`. This is mostly so that the tests are a bit easier. IMHO we could get rid of `sorted` and just apply `order(id: :desc) to the few places where we don't override it
